File: src/merge_sort.py
import random


def msort(arr):
    temp = [None] * len(arr)
    msort_impl(arr, 0, len(arr), temp)


def msort_impl(arr, begin, end, temp):
    # 예를 들어 [0, 1]이면
    # begin=0, end = 2 (end - begin = len이 되도록)
    arr_len = end - begin

    # 기저 상태
    if arr_len <= 1:
        return

    mid = begin + (arr_len // 2)
    msort_impl(arr, begin, mid, temp)
    msort_impl(arr, mid, end, temp)
    # print(f"msort: {begin}..{mid}..{end} ({arr_len})")

    # 결국 mergesort가 selection sort보다 효율적인 이유에는 **이미 정렬된 두 배열이 존재한다**는 조건이 붙으면
    # 병합한 배열도 정렬되게 만들 수 있는 효율적인 방법이 존재한다는 거겠지.
    offset = begin
    left_offset = begin
    right_offset = mid
    while left_offset < mid and right_offset < end:
        if arr[left_offset] < arr[right_offset]:
            temp[offset] = arr[left_offset]
            left_offset += 1
        else:
            temp[offset] = arr[right_offset]
            right_offset += 1

        offset += 1

    # 과연, if문이 4개 들어가서 복사하는 것보다 (left가 mid를 넘어갔는가, right가 end를 넘어갔는가, left <=> right 비교)
    # left, right 조건은 나중에 비교하고 한번에 memcpy()로 복사하는 게 더 빠르겠지. 이쪽은 많은 언어에서 simd 지원도 받고..
    if left_offset == mid:
        temp[offset:end] = arr[right_offset:end]
    else:
        temp[offset:end] = arr[left_offset:mid]

    # 복사
    arr[begin:end] = temp[begin:end]


def main():
    data = [4, 3, 42, 1, 2, 5, 6, 7]
    msort(data)
    print(data)

    # 홀수
    data = [4, 3, 42, 1, 2, 5, 7]
    msort(data)
    print(data)

    print("====100====")
    data = [random.randint(0, 10000) for _ in range(100)]
    msort(data)
    print(data)

    print("====sorting 1m elements====")
    data = [random.randint(0, 1000000) for _ in range(1_000_000)]
    msort(data)
    print("testing..")
    # 이쯤되면 눈으로 판단하는 게 어려우니
    m = -1
    for val in data:
        m = max(m, val)
        if m != val:
            print(f"something is wrong ({m} > {val})")
    print("done!")


if __name__ == "__main__":
    main()

File: src/merge_sort.cpp
#include <spdlog/spdlog.h>
#include <vector>
#include <limits>
#include <algorithm>
#include <random>

void merge_sort_impl(std::vector<int>& arr, size_t begin, size_t end, std::vector<int>& temp) {
    auto arr_len = end - begin;

    if (arr_len <= 1)
        return;

    auto mid = begin + (arr_len / 2);
    merge_sort_impl(arr, begin, mid, temp);
    merge_sort_impl(arr, mid, end, temp);

    // merge(left, right)
    // iterator를 받고 증가시키면 밑에서 복사할 때 더 짧게 쓸 수 있곘다는 생각은 들지만...
    auto offset = begin;
    auto left = begin;
    auto right = mid;

    while (left < mid && right < end) {
        if (arr[left] < arr[right]) {
            temp[offset] = arr[left];
            left += 1;
        } else {
            temp[offset] = arr[right];
            right += 1;
        }

        offset += 1;
    }

    if (left == mid)
        std::copy(arr.begin() + right, arr.begin() + end, temp.begin() + offset);
    else
        std::copy(arr.begin() + left, arr.begin() + mid, temp.begin() + offset);

    std::copy(temp.begin() + begin, temp.begin() + end, arr.begin() + begin);
}

void merge_sort(std::vector<int>& arr) {
    std::vector<int> temp(arr.size(), 0);
    merge_sort_impl(arr, 0, arr.size(), temp);
}

void check_sorted(const std::vector<int>& arr) {
    spdlog::info("checking {} elements", arr.size());

    int m = std::numeric_limits<int>::min();

    for (auto val : arr) {
        if (m > val) {
            spdlog::error("\u274C found an incorrectly sorted element: m={}, val={}", m, val);
            return;
        }

        m = val;
    }

    spdlog::info("\u2705 passed");
}

int main() {
    // int 말고도 generic한 타입 아무거나 괜찮지만...
    // 역시 template<typename T> 계속 넣기도 번거롭고
    // 제일 중요한 건 오류가 발생하면 다른 언어의 generic 기능에 비해 오류 메시지를 읽기가 상당히 골룸하단 말이지...

    std::vector<int> data;

    data = { 0, 4, 3, 7, 6, 2 };
    merge_sort(data);
    check_sorted(data);

    // ... 500만개를 랜덤으로 채우고 정렬
    data = std::vector(5e6, 0);
    std::random_device rd;
    std::mt19937 rng(rd());
    std::uniform_int_distribution<> uniform(-1e6);
    for (auto i = 0; i < 1e6; i++) {
        data[i] = uniform(rng);
    }

    //    [2021-05-14 15:33:44.028] [info] sorting 5000000 elements..
    //    [2021-05-14 15:33:44.717] [info] done!
    //    [2021-05-14 15:33:44.718] [info] checking 5000000 elements
    //    [2021-05-14 15:33:44.722] [info] ✅ passed
    // 1초도 안 걸린다!
    spdlog::info("sorting {} elements..", data.size());
    merge_sort(data);
    spdlog::info("done!");
    check_sorted(data);

    return 0;
}
